### PR TITLE
Ntfs path hint

### PR DESCRIPTION
**Pull request by stbe**
## Description:
Adds a `path_hint` attribute to event data that gets constructed from walking each entries `parent_file_reference`.

This should be properly reviewed as it adds runtime and memory consumption. Also, the approach taken might need to be discussed if someone sees problems with it. The MFT gets parsed twice now: this PR adds a round of MFT parsing to collect `file_reference`, `name`, and the `parent_reference` in a dictionary. The second round in which all other attributes get extracted then populates the path hint by resolving all parents recursively from that dictionary

**Related issue (if applicable):** adds to #316

## Notes:
All contributions to Plaso undergo [code
review](https://github.com/log2timeline/l2tdocs/blob/master/process/Code%20review%20process.asciidoc). This makes sure
that the code has appropriate test coverage and conforms to the Plaso [style
guide](https://plaso.readthedocs.io/en/latest/sources/developer/Style-guide.html).

One of the maintainers will examine your code, and may request changes. Check off the items below in
order, and then a maintainer will review your code.

## Checklist:
  - [x] Automated checks (Travis, Codecov, Codefactor )pass
  - [x] No new [new dependencies](https://plaso.readthedocs.io/en/latest/sources/developer/Adding-a-new-dependency.html) are required or l2tdevtools has been updated
  - [x] Reviewer assigned

### CODE REVIEW

**Comment by stbe** [comment]
Sure @Onager! But I just realized the the changes need improvement first. There is currently two issues I see with the implementation:
1) does not account for things like hard links: implementation currently takes the last parent it gets from iterating the entries attributes (there might be multiple $FN in case of a hardlink, each having a different parent)
2) writes a path hint for events generated by the $SI attribute (related to 1)

File: tests/parsers/ntfs.py [hunk 1]
```diff
@@ -57,6 +57,60 @@ def testParseFile(self):
...
+    self.assertEqual(event_data.name, 'SAM')
+    self.assertEqual(event_data.path_hint, './WINDOWS/system32/config/SAM')
+
+    # Test path_hint with deleted file 'CAJA1S19.js'
+    event = events[120480]
+    event_data = self._GetEventDataOfEvent(storage_writer, event)
+    self.assertEqual(event_data.name, 'CAJA1S19.js')
+    self.assertEqual(
+        event_data.path_hint,
+        './Documents and Settings/Donald Blake/Local Settings/Temporary Internet Files/Content.IE5/9EUWFPZ1/CAJA1S19.js')  # pylint: disable=line-too-long
```
Comment (joachimmetz) [inline tests/parsers/ntfs.py]: Please don't do these style overrides there is a style guide for a reason.

File: plaso/parsers/ntfs.py [hunk 2]
```diff
@@ -22,6 +22,7 @@
 from plaso.parsers import manager
 
 
+# pylint: disable=too-many-instance-attributes,too-few-public-methods
```
Comment (joachimmetz) [inline plaso/parsers/ntfs.py]: which version of pylint did you use?

File: plaso/parsers/ntfs.py [hunk 3]
```diff
@@ -90,6 +94,14 @@ class NTFSMFTParser(interface.FileObjectParser):
...
   _MFT_ATTRIBUTE_FILE_NAME = 0x00000030
   _MFT_ATTRIBUTE_OBJECT_ID = 0x00000040
+  _PATH_SEPARATOR = '/'
+  _PATH_NO_NAME_REPLACEMENT = '???'
+  _PATH_NAME_ORPHAN = '$Orphan'
+
+  def __init__(self):
+    """Intializes the NTFS MFT Parser"""
+    super(NTFSMFTParser, self).__init__()
+    self.path_info = dict()
```
Comment (joachimmetz) [inline plaso/parsers/ntfs.py]: why not `self.path_info = {}` ?

File: plaso/parsers/ntfs.py [hunk 4]
```diff
@@ -275,6 +300,238 @@ def _ParseMFTEntry(self, parser_mediator, mft_entry):
...
+    Returns:
+      tuple: A tuple of (name, parent_record_number, parent_sequence_number)
+
+    """
+
+    name = None
+    parent_record_number = None
+    parent_sequence_number = None
+
+    # Sort the attributes by re-mapping the `namespace` values
```
Comment (joachimmetz) [inline plaso/parsers/ntfs.py]: code is not use, please remove

File: plaso/parsers/ntfs.py [hunk 5]
```diff
@@ -275,6 +300,238 @@ def _ParseMFTEntry(self, parser_mediator, mft_entry):
...
+    # ns_map = {
+    #   0x0: 0x2,
+    #   0x1: 0x1,
+    #   0x2: 0x3,
+    #   0x3: 0x0
+    # }
+    # file_name_attributes.sort(key=lambda a: (ns_map[a[4]], a[1]))
+
+    # Sort by name length and attribute index
+    attribute_infos.sort(key=lambda a: (len(a[0]), a[1]), reverse=True)
```
Comment (joachimmetz) [inline plaso/parsers/ntfs.py]: per style guide `lambda a` => `lambda attribute`

File: plaso/parsers/ntfs.py [hunk 6]
```diff
@@ -90,6 +94,14 @@ class NTFSMFTParser(interface.FileObjectParser):
   _MFT_ATTRIBUTE_STANDARD_INFORMATION = 0x00000010
   _MFT_ATTRIBUTE_FILE_NAME = 0x00000030
   _MFT_ATTRIBUTE_OBJECT_ID = 0x00000040
+  _PATH_SEPARATOR = '/'
```
Comment (joachimmetz) [inline plaso/parsers/ntfs.py]: Why use the forward slash?

File: plaso/parsers/ntfs.py [hunk 7]
```diff
@@ -291,6 +548,16 @@ def ParseFileObject(self, parser_mediator, file_object):
       parser_mediator.ProduceExtractionWarning(
           'unable to open file with error: {0!s}'.format(exception))
 
+    # Collect path information in a first round of parsing
+    for entry_index in range(0, mft_metadata_file.number_of_file_entries):
+      try:
+        mft_entry = mft_metadata_file.get_file_entry(entry_index)
+        self._CollectMFTEntryPathInfo(mft_entry)
```
Comment (joachimmetz) [inline plaso/parsers/ntfs.py]: doing this for every MFT entry can be memory intensive. Technically we only need to cache the parent path hints in `self.path_info`

**Comment by joachimmetz** [comment]
@stbe sry for the delay on my side. I think the outline is good but there are a couple of caveats with the implementation the ones you highlight, memory usage and having to iterate all MFT entries twice.

**Comment by joachimmetz** [comment]
I made some changes to libfsntfs to support a path hint per `$FILE_NAME` attribute

And made some changes to the mft parser in https://github.com/log2timeline/plaso/pull/2780. This implementation:
* accounts for hardlinks and names in the DOS namespace
* multiple path hints are returned for `$STANDARD_INFORMATION` attribute when appropriate
* parent directory names in the DOS namespace are ignored for now, since they are synonymous to those in the Windows namespace
* iterates all the MFT entries only once
* only the parent path hints are cached

**Comment by joachimmetz** [comment]
Closed in favor of https://github.com/log2timeline/plaso/pull/2780

**Pull request closed without merging**